File: javascript/src/one_to_n.ts
// Printing 1 to n iteratively and recursively
import { Trampoline, tramp } from "./global";
function oneToN(n: number): void {
    for (let idx = 0; idx < n; idx++) {
        console.log(idx);
    }
}

function oneToNTailRec(n: number): void {
    function helper(idx: number): void {
        if (idx == n) {
            return;
        }
        console.log(idx);
        return helper(idx + 1);
    }
    return helper(0);
}

function oneToNThunk(n: number): Trampoline<void> {
    function helper(idx: number): Trampoline<void> {
        if (idx == n) {
            return;
        }
        console.log(idx);
        return () => helper(idx + 1);
    }
    return helper(0);
}

oneToN(10);
oneToNTailRec(10);
tramp(oneToNThunk, 100_000);

File: python/src/one_to_n.py
# Printing 1 to n iteratively and recursively
from common import time_execution


@time_execution()
def one_to_n(n) -> list[int]:
    result = []

    def helper(start: int = 0) -> list[int]:
        if start > n:
            return
        result.append(start)
        return helper(start+1)
    helper()
    return result


@time_execution(isTrampoline=True)
def one_to_n_gen(n) -> list[int]:

    def helper(start=0, res=[]) -> list[int]:
        if start > n:
            yield res
        yield helper(start + 1, res+[start])
    return helper()


print(one_to_n(10))
# print(one_to_n(10000))
print(one_to_n_gen(10_000))
